Merge 'DTCs' branch into 'develop'

DynamicEwok opened the pull request:
Contains implementation of service 0x14 and 0x19 sub-functions to interface with the DHM's DTCs.

Service 0x19 sub-functions 0x4 and 0x6 and Service 0x85 will be added in a future update.

cmcerove committed:
Work in progress for adding DTC services to uds.py

cmcerove committed:
Merge branch 'develop' into DTCs

Commit:
Implemented support for service 0x19 sub-functions 0x2, 0x7-0xA, 0xC, and 0xE.

Commit:
Now using uds._check to build/validate request before sending to DHM.

Altered snipping return list to not cut out portions of the return message.

Commit:
Added failure case where _check receives incompatible variable type.

Commit:
Error is raised when attempting to call an unimplemented sub-function for Service 0x19

Commit:
Implemented handling of Service 0x19 sub-functions 0x4 and 0x6

Commit:
Fixed copy/paste error causing subfunction 0x04 to be called instead of 0x06.

cmcerove commented on pyvxl/uds.py:
```diff
@@ -207,13 +218,319 @@ def ecu_reset(self, reset_type, raise_error=True, **kwargs):
+        result = None
+
+        sub_function_byte = self._check('sub_function', sub_function)
+        match(sub_function_byte[0]):
```
```suggestion
        match sub_function_byte[0]:
```

cmcerove commented on pyvxl/uds.py:
```diff
@@ -207,13 +218,319 @@ def ecu_reset(self, reset_type, raise_error=True, **kwargs):
+
+        return result
+
+    def read_dtc_snapshot_identification(self):
```
```suggestion
    def read_dtc_snapshot_identification(self, *args, **kwargs):
```
So someone calling this reaches the error within rather than finding an issue with an argument they're passing in first.

Commit:
Refactored read DTC functions to have less repetitive code by using a helper function.

Could still improve more by not having separate calls for each sub-function, and instead build the request in the read_dtc_info function and call the helper from there. Everything about calling the sub-function is uniform outside of building the request and trimming to response to not include useless info.

cmcerove commented on pyvxl/uds.py:
```diff
@@ -207,13 +218,257 @@ def ecu_reset(self, reset_type, raise_error=True, **kwargs):
         return result
 
-    def clear_dtcs(self, *args, **kwargs):
+    def clear_dtc_info(self, dtc_group, raise_error=True, **kwargs):
```
Could you update the docstrings to follow google style? It looks like most of them are but a few were missed like this function

Commit:
Updated docstrings and some incorrect syntax.
